### ISSUE TITLE
AES-256: srtp_protect_rtcp does not check for SRTP_AES_256_ICM, wrong nonce is used

### ISSUE THREAD

**Issue by aalba6675** [cisco/libsrtp#264]
If OpenSSL is chosen as provider of AES ICM, then the cipher id is set to SRTP_AES_256_ICM, this mean an incorrect nonce code path is chosen for SRTCP. (SRTP is not affected)

2.1.0-pre, srtp/srtp.c
Using OpenSSL the cipher id is set to SRTP_AES_256_ICM
(Without OpenSSL the cipher id is set to SRTP_AES_ICM)

In `srtp_protect()` and `srtp_unprotect()` a check is made for both SRTP_AES_ICM SRTP_AES_256_ICM in initialising the ICM nonce:
```c
  if (session_keys->rtp_cipher->type->id == SRTP_AES_ICM ||
       session_keys->rtp_cipher->type->id == SRTP_AES_256_ICM) {
```

In `srtp_protect_rtcp()` and `srtp_unprotect_rtcp()` the same check is **not made**   
```c
  /*  NO check for SRTP_AES_256_ICM
   * if we're using rindael counter mode, set nonce and seq 
   */
  if (session_keys->rtcp_cipher->type->id == SRTP_AES_ICM) {
    v128_t iv;
  ```

This violates RFC 6188; also it will lead to keystream reuse across consecutive RTCP packets in AES-256.

(Both rtp/rtcp protect unprotect functions will not work for AES-192 as there is no check for SRTP_AES_192_ICM)

Fix: set all cipher ids to SRTP_AES_ICM with OpenSSL support enabled?
This probably affects all 1.5.x versions as well.

**Comment by rchan1611** [comment]
+1 Hit this when testing FreeSWITCH 1.6.15, which uses an in-tree libsrtp 1.4.5, and distributes packages with OpenSSL support compiled in. 

The SRTCP packets are non-RFC6188 compliant (wrong counter used)

**Comment by pabuhler** [comment]
Hi @rchan1611,

I think this will be fixed if #253 is merged in. If you see: https://github.com/pabuhler/libsrtp/blob/d335bd88b147cf0fdf956bd21a55d3c7c9d974d7/srtp/srtp.c#L4032  and https://github.com/pabuhler/libsrtp/blob/d335bd88b147cf0fdf956bd21a55d3c7c9d974d7/srtp/srtp.c#L3773  .
Could you confirm it will fix it for you?
Will consider back porting your fix in #266 to the 1.5.x branch as it is definitely a problem.

**Comment by aalba6675** [comment]
Yes my use case is covered by #253. For 1.5.x something similar is needed. Please close as "duplicate" unless you want to leave it open until something is done for 1.5.x.

Thank you!

**Comment by pabuhler** [comment]
Leaving open, #266 should be applied to 1.5.x, maybe with some tests ....

**Comment by paulej** [comment]
As the number of fixes for 1.5 mount, it makes me concerned that we're dealing with two distinct releases.  What keeps people from moving to 2.1?  2.0 is considered stable and has been out for a while.  It would be good if folks could put that through its paces if concerned about going from 1.5 to 2.1 directly. (Note that 2.0 also has errors in the GCM logic that were recently discovered.)

**Comment by mjerris** [comment]
as soon as 2.1 is out, i'll take freeswitch to it for our next major version.  We need this and the MKI features.

**Comment by paulej** [comment]
You're the brave one. :-)  Actually, I generally feel pretty confident with the libsrtp since, when it doesn't work... it's often quite painfully obvious.  That said, there are hidden bugs like the AES GCM RTCP IV issue.  That will be a bit painful.

**Comment by pabuhler** [comment]
@paulej Yes it looks like two releases. I think though that it is important that the latest 1.5.x and the latest 2.x.x are compatible, and now there looks like at least two changes that will break compatibility. Perhaps it would even be better to make it a 1.6.0 so it is explicit.

**Comment by rchan1611** [comment]
+1 as @aalba6675, using 1.x and need the fix for AES-256.

**Comment by pabuhler** [comment]
Closing as now fixed in upcoming 2.1 & 1.6 releases
